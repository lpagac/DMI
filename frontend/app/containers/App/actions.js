import {
  LOAD_STRINGS,
  LOAD_STRINGS_SUCCESS,
  LOAD_STRINGS_ERROR,
  ADD_STRING,
  ADD_STRING_ERROR,
  ADD_STRING_SUCCESS,
  CLEAR_ADD_NOTIFICATIONS,
} from './constants';

/**
 * Load the strings, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_STRINGS
 */
export function loadStrings() {
  return {
    type: LOAD_STRINGS,
  };
}

/**
 * Dispatched when the strings are loaded by the request saga
 *
 * @param  {array} strings array of strings from db
 *
 * @return {object}      An action object with a type of LOAD_STRINGS_SUCCESS passing the strings
 */
export function stringsLoaded(strings) {
  return {
    type: LOAD_STRINGS_SUCCESS,
    strings,
  };
}

/**
 * Dispatched when loading the strings fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_STRINGS_ERROR passing the error
 */
export function stringsLoadingError(error) {
  return {
    type: LOAD_STRINGS_ERROR,
    error,
  };
}

/**
 * Add a new string, this action starts the request saga
 *
 * @param {string} id the uuid for the string to add
 *
 * @param {string} string the string to add
 *
 * @return {object} An action object with a type of ADD_STRING, passes string and id
 */

export function addString(id, string) {
  return {
    type: ADD_STRING,
    id,
    string,
  };
}

/**
 * Dispatched when new string is added succesfully
 *
 * @param {string} id the uuid for the string that was added
 *
 * @param {string} string the string that was added
 *
 * @param {string} strings the strings that were returned from the saga
 *
 * @return {object} An action object with a type of ADD_STRING_SUCCESS, passes string and id and strings
 */
export function stringAdded(id, string, strings = false) {
  return {
    type: ADD_STRING_SUCCESS,
    id,
    string,
    strings,
  };
}

/**
 * Dispatched when an error occurs when trying to post a string
 *
 * @param  {object} error The error
 *
 * @return {object} An action object with a type of ADD_STRING_ERROR, passes the error message
 */

export function stringAddingError(error) {
  return {
    type: ADD_STRING_ERROR,
    error,
  };
}

/**
 * Dispatched when an error success occurs when trying to post a string
 *
 * @param  {object}  The error
 *
 * @return {object} An action object with a type of ADD_STRING_ERROR, passes the error message
 */

export function clearNotifications() {
  return {
    type: CLEAR_ADD_NOTIFICATIONS,
  };
}
