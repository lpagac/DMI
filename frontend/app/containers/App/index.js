/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 */

import React from 'react';
import { Helmet } from 'react-helmet';
import { Switch, Route, Redirect } from 'react-router-dom';
import styled from 'styled-components';

import HomePage from 'containers/HomePage/Loadable';
import NavBar from 'components/NavBar';
import StringsPage from 'containers/StringsPage/Loadable';
import AddStringPage from 'containers/AddStringPage/Loadable';

import GlobalStyle from '../../global-styles';

const AppWrapper = styled.div`
  display: flex;
  width: 100%;
  flex-direction: column;
  align-items: center;
  margin: auto;
  padding: 40px 0;
`;

export default function App() {
  return (
    <div>
      <NavBar />
      <AppWrapper>
        <Helmet
          titleTemplate="Strings App Take Home"
          defaultTitle="Strings App"
        >
          <meta name="description" content="A Strings app take home." />
        </Helmet>
        <Switch>
          <Route exact path="/" component={HomePage} />
          <Route exact path="/strings" component={StringsPage} />
          <Route exact path="/strings/add" component={AddStringPage} />
          <Redirect to="/" />
        </Switch>
        <GlobalStyle />
      </AppWrapper>
    </div>
  );
}
