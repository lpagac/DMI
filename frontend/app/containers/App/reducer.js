/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import {
  LOAD_STRINGS_SUCCESS,
  LOAD_STRINGS,
  LOAD_STRINGS_ERROR,
  ADD_STRING,
  ADD_STRING_SUCCESS,
  ADD_STRING_ERROR,
  CLEAR_ADD_NOTIFICATIONS,
} from './constants';

// The initial state of the App
export const initialState = {
  loading: false,
  error: false,
  strings: false,
  addingString: false,
  addingStringSuccess: false,
  addingStringError: false,
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOAD_STRINGS: {
        draft.loading = true;
        draft.error = false;
        draft.strings = false;
        break;
      }

      case LOAD_STRINGS_SUCCESS: {
        draft.strings = action.strings;
        draft.loading = false;
        draft.error = false;
        break;
      }

      case LOAD_STRINGS_ERROR: {
        draft.error = action.error;
        draft.loading = false;
        draft.strings = false;
        break;
      }

      case ADD_STRING: {
        draft.addingString = true;
        draft.addingStringError = false;
        draft.addingStringSuccess = false;
        break;
      }

      case ADD_STRING_SUCCESS: {
        const { id, string } = action;
        if (draft.strings) {
          draft.strings.unshift({ id, string });
        } else {
          draft.strings = action.strings;
        }
        draft.addingStringSuccess = true;
        draft.addingString = false;
        draft.addingStringError = false;
        break;
      }
      case ADD_STRING_ERROR: {
        draft.addingStringError = action.error;
        draft.addingString = false;
        draft.addingStringSuccess = false;
        break;
      }
      case CLEAR_ADD_NOTIFICATIONS: {
        draft.addingStringError = false;
        draft.addingString = false;
        draft.addingStringSuccess = false;
        break;
      }
    }
  });

export default appReducer;
