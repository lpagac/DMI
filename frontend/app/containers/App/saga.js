/**
 * Handles fetching all strings and adding new string
 */

import {
  put,
  call,
  takeLatest,
  takeEvery,
  all,
  delay,
} from 'redux-saga/effects';
import {
  LOAD_STRINGS,
  ADD_STRING,
  ADD_STRING_SUCCESS,
  ADD_STRING_ERROR,
} from 'containers/App/constants';

import {
  stringsLoaded,
  stringsLoadingError,
  stringAdded,
  stringAddingError,
  clearNotifications,
} from 'containers/App/actions';

import request from 'utils/request';

const BASE_API_URL = `http://localhost:3001/api/strings`;

/**
 * strings list request/response handler
 */
export function* getStrings() {
  try {
    const resp = yield call(request, BASE_API_URL);
    yield put(stringsLoaded(resp.strings));
  } catch (err) {
    yield put(stringsLoadingError(err));
  }
}

/**
 * watcher for LOAD_STRINGS actions
 */
export function* watchLoadStrings() {
  yield takeLatest(LOAD_STRINGS, getStrings);
}

/** adds new string via post request to server */

export function* addString(action) {
  try {
    const { id, string } = action;
    const options = {
      method: 'POST',
      cache: 'no-cache',
      mode: 'cors',
      headers: {
        'Content-Type': 'application/json',
      },
      redirect: 'error',
      body: JSON.stringify({ id, string }),
    };
    const resp = yield call(request, BASE_API_URL, options);
    yield put(stringAdded(id, string, resp.strings));
  } catch (err) {
    yield put(stringAddingError(err.message));
  }
}

/** post new string watcher */

export function* watchAddString() {
  yield takeEvery(ADD_STRING, addString);
}

/** clears adding string notifications */

export function* clearUserNotifications() {
  yield delay(3000);
  yield put(clearNotifications());
}

/** watch for result of addString (success/error) */

export function* watchAddStringNotifications() {
  yield all([
    takeLatest(ADD_STRING_SUCCESS, clearUserNotifications),
    takeLatest(ADD_STRING_ERROR, clearUserNotifications),
  ]);
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* rootSaga() {
  yield all([
    watchLoadStrings(),
    watchAddString(),
    watchAddStringNotifications(),
  ]);
}
