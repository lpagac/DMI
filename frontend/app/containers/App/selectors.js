import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectRouter = state => state.router;
const selectGlobal = state => state.global || initialState;

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectStrings = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.strings,
  );

const makeSelectAddingString = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.addingString,
  );
const makeSelectAddingStringSuccess = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.addingStringSuccess,
  );
const makeSelectAddingStringError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.addingStringError,
  );

const makeSelectLocation = () =>
  createSelector(
    selectRouter,
    routerState => routerState.location,
  );

export {
  makeSelectLoading,
  makeSelectError,
  makeSelectStrings,
  makeSelectAddingString,
  makeSelectAddingStringSuccess,
  makeSelectAddingStringError,
  makeSelectLocation,
};
