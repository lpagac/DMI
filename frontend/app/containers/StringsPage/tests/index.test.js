/**
 *
 * Tests for StringsPage
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { Provider } from 'react-redux';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { browserHistory } from 'react-router-dom';
// import 'jest-dom/extend-expect'; // add some helpful assertions

import { loadStrings } from '../../App/actions';
import { StringsPage, mapDispatchToProps } from '../index';
import { DEFAULT_LOCALE } from '../../../i18n';
import configureStore from '../../../configureStore';

describe('<StringsPage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const fetchMock = jest.fn();
    render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <StringsPage
            loading={false}
            error={false}
            strings={false}
            fetchStrings={fetchMock}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const fetchMock = jest.fn();
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <StringsPage
            loading={false}
            error={false}
            strings={false}
            fetchStrings={fetchMock}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  describe('mapDispatchToProps', () => {
    describe('fetchStrings', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.fetchStrings).toBeDefined();
      });

      it('should dispatch loadStrings when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        result.fetchStrings();
        expect(dispatch).toHaveBeenCalledWith(loadStrings());
      });
    });
  });

  describe('<StringsSection/>', () => {
    it('should display current strings', () => {
      const fetchMock = jest.fn();
      const utils = render(
        <Provider store={store}>
          <IntlProvider locale={DEFAULT_LOCALE}>
            <StringsPage
              loading={false}
              error={false}
              strings={[{ id: 'test-id', string: 'test string' }]}
              fetchStrings={fetchMock}
            />
          </IntlProvider>
        </Provider>,
      );

      const stringsSection = utils.getByText('test string');
      expect(stringsSection).not.toEqual(null);
    });
  });
});
