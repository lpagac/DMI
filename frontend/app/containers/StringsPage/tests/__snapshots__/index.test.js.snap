// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StringsPage /> Should render and match the snapshot 1`] = `
<div>
  <section
    class="CenteredSection-sc-1iimrbg-0 hHmbwO"
  >
    <h1
      class="H1-tncdg6-0 kWlXoI"
    >
      <span>
        Pretty sweet collection of strings right?
      </span>
    </h1>
  </section>
</div>
`;
